include/pfn/expected.hpp

```diff
@@ -0,0 +1,1483 @@
+      -> decltype(_transform(*this, FWD(f)))
+  {
+    return _transform(*this, FWD(f));
+  }
+  template <class F>
+  constexpr auto transform(F &&f) const &&                       //
+      noexcept(noexcept(_transform(::std::move(*this), FWD(f)))) // extension
+      -> decltype(_transform(::std::move(*this), FWD(f)))
+  {
+    return _transform(::std::move(*this), FWD(f));
+  }
+
+  template <class F>
+  constexpr auto transform_error(F &&f) &                 //
+      noexcept(noexcept(_transform_error(*this, FWD(f)))) // extension
+      -> decltype(_transform_error(*this, FWD(f)))
+  {
+    return _transform_error(*this, FWD(f));
+  }
+  template <class F>
+  constexpr auto transform_error(F &&f) &&                             //
+      noexcept(noexcept(_transform_error(::std::move(*this), FWD(f)))) // extension
+      -> decltype(_transform_error(::std::move(*this), FWD(f)))
+  {
+    return _transform_error(::std::move(*this), FWD(f));
+  }
+  template <class F>
+  constexpr auto transform_error(F &&f) const &           //
+      noexcept(noexcept(_transform_error(*this, FWD(f)))) // extension
+      -> decltype(_transform_error(*this, FWD(f)))
+  {
+    return _transform_error(*this, FWD(f));
+  }
+  template <class F>
+  constexpr auto transform_error(F &&f) const &&                       //
+      noexcept(noexcept(_transform_error(::std::move(*this), FWD(f)))) // extension
+      -> decltype(_transform_error(::std::move(*this), FWD(f)))
+  {
+    return _transform_error(::std::move(*this), FWD(f));
+  }
+
+  // [expected.object.eq], equality operators
+  template <class T2, class E2>
+    requires(not ::std::is_void_v<T2>)
+  constexpr friend bool operator==(expected const &x, expected<T2, E2> const &y) //
+      noexcept(noexcept(*x == *y) && noexcept(x.error() == y.error()))           // extension
+    requires ( //
+requires {
+      { *x == *y } -> std::convertible_to<bool>;
+    } &&
+requires {
+    { x.error() == y.error() } -> std::convertible_to<bool>;
+    })
+  {
+    if (x.has_value() != y.has_value()) {
+      return false;
+    } else if (x.has_value()) {
+      return *x == *y;
+    } else {
+      return x.error() == y.error();
+    }
+  }
+  template <class T2>
+    requires(not detail::_is_some_expected<T2>)
+  constexpr friend bool operator==(expected const &x, const T2 &v) //
+      noexcept(noexcept(*x == v))                                  // extension
+    requires requires {
+      { *x == v } -> std::convertible_to<bool>;
+    }
+  {
+    return x.has_value() && static_cast<bool>(*x == v);
+  }
+  template <class E2>
+  constexpr friend bool operator==(expected const &x, unexpected<E2> const &e) //
+      noexcept(noexcept(x.error() == e.error()))                               // extension
+    requires requires {
+      { x.error() == e.error() } -> std::convertible_to<bool>;
+    }
+  {
+    return (not x.has_value()) && static_cast<bool>(x.error() == e.error());
+  }
+
+private:
+  union {
```

Comment:
Would it be better to use a std::variant? I don't think it's going to be slower.

Reply:
... slower, at the very least in the sense of compilation speed. Likely in runtime, too (because of prerequisites checks).

Reply:
I think it [shouldn't have any impact](https://godbolt.org/z/hMPjMjqWh) in the runtime but yeah I think it'll introduce <variant> which is not relevant.